src/workqueue/workqueue.c

```diff
@@ -0,0 +1,849 @@
+
+// Called by the management system when it wants to inject a
+// workitem that will case a worker thread to wait
+static void _dispatch_worker_should_wait(void);
+
+// Called by the management system when it wants to inject a
+// workitem that will case a worker thread to terminate
+static void _dispatch_worker_should_exit(void);
+
+// Called by the management system when it wants to
+// create an additional worker thread.
+static void _dispatch_worker_should_start(void);
+
+#if DISPATCH_ENABLE_PWQ_KEXT
+//////////////////////////////////////////////////////////////////
+// versions of above functions required for the kernel extension
+// temporary to isolate kext code to separate file.
+//////////////////////////////////////////////////////////////////
+
+int _dispatch_workqueue_management_init_kext(void);
+void _dispatch_worker_register_kext(void);
+void _dispatch_worker_unregister_kext(void);
+void _dispatch_work_added_kext(void);
+void _dispatch_worker_wait_kext(void);
+void* dispatch_thread_monitor_main_kext(void *context DISPATCH_UNUSED);
+
+#endif // DISPATCH_ENABLE_PWQ_KEXT
+
+/////////////////////////////////
+// Workqueue internal data structures and functions
+/////////////////////////////////
+
+// Debug messages for threadpool operations
+static bool threadpool_verbose;
+#define verbose_printf(fmt,...) do {								\
+	if (unlikely(threadpool_verbose))								\
+		fprintf(stderr, "libdispatch_threadpool: "fmt"\n", ##__VA_ARGS__); \
+} while (0)
+
+struct work {
+	STAILQ_ENTRY(work) item_entry;
+	void (*func)(void *);
+	void *func_arg;
+};
+typedef struct work * dispatch_workitem_t;
+
+typedef struct _dispatch_workqueue {
+	STAILQ_HEAD(,work) item_listhead;
+	pthread_spinlock_t sl;
+	int priority;
+	bool overcommit;
+} dispatch_workqueue_s;
+typedef dispatch_workqueue_s *dispatch_workqueue_t;
+
+/*
+ * Overcommit queues.
+ *
+ * The overcommit pool uses a simple coarse locking policy:
+ *   all overcommit pool operations are performed while
+ *   holding the mutex.
+ *
+ * This enables a simple, but non-scalable, implementation
+ * that is adequate as long as operations on the overcommit
+ * pool are relatively infrequent.
+ */
+typedef struct {
+	uint mask;
+	int num_spares;
+	dispatch_workqueue_t queues[WORKQ_NUM_PRIORITIES];
+	pthread_mutex_t mutex;
+	pthread_cond_t spare_workers;
+} dispatch_overcommit_pool_s;
+typedef dispatch_overcommit_pool_s * dispatch_overcommit_pool_t;
+
+static dispatch_overcommit_pool_s overcommit_pool;
+
+/*
+ * Normal queues.
+ *
+ * The normal pool uses a combination of atomic
+ * operations on a volatile mask word that encodes which
+ * workqueues are non-empty and per-queue synchronization
+ * (currently implemented via pthread_spin_locks).
+ * This is slightly more scalable than the overcommit queue, but
+ * is may be an area that needs future performance enhancement.
+ */
+typedef struct {
+	volatile uint mask;
+	dispatch_workqueue_t queues[WORKQ_NUM_PRIORITIES];
```

Comment:
work *queues* is confusing, this is really about managing a thread pool for a given priority bucket.

I think the whole type should rather be named:

`dispatch_wq_thread_pool_t` and be referred as a `thread_pool`. or somesuch (just a suggestion, but queue is so overloaded in dispatch that I would shy away from using the term even more)